File: leetcode/easy/637-average-levels/iterative.py
"""
  @author Vortexx2
  Problem 637

  Runtime - 40 ms
  Memory Usage - 16.6 MB
"""
from typing import List, Optional


class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
    res = []
    level = [root]

    while level:
      nextLevel, currLevel = [], []
      currSum = 0

      for node in level:
        currSum += node.val
        if node.left:
          nextLevel.append(node.left)

        if node.right:
          nextLevel.append(node.right)

      res.append(currSum / len(level))
      level = nextLevel

    return res

File: leetcode/easy/637-average-levels/iterative.cpp
/**
 * @file stack.cpp
 * @author Vortexx2
 * @brief Problem 637
 * @date 21-08-2021
 * 
 * Runtime - 18 ms
 * Memory Usage - 22.9 MB
 */
#include <algorithm>
#include <climits>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<double> averageOfLevels(TreeNode *root) {
    vector<double> res;
    vector<TreeNode *> level = {root};

    while (!level.empty()) {
      vector<int> currentLevel;
      vector<TreeNode *> nextLevel;
      long sum = 0;

      for (auto &node : level) {
        sum += node->val;
        if (node->left) nextLevel.push_back(node->left);
        if (node->right) nextLevel.push_back(node->right);
      }

      res.push_back((double)(sum) / level.size());
      level = nextLevel;
    }

    return res;
  }
};